src/wallet/wallet.cpp

```diff
@@ -5569,3 +5851,72 @@ KeyAddResult AddSpendingKeyToWallet::operator()(const libzcash::SaplingExtendedS
 KeyAddResult AddSpendingKeyToWallet::operator()(const libzcash::InvalidEncoding& no) const {
     throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid spending key");
 }
+
+std::optional<libzcash::UnifiedAddress> LookupUnifiedAddress::operator()(const libzcash::SaplingPaymentAddress& saplingAddr) const {
+    auto ufvkPair = wallet.GetUFVKMetadataForReceiver(saplingAddr);
+    if (ufvkPair.has_value()) {
+        auto ufvkid = ufvkPair.value().first;
+        auto ufvk = wallet.GetUnifiedFullViewingKey(ufvkid);
+        if (!(ufvk.has_value() && ufvk.value().GetSaplingKey().has_value())) {
+            throw std::runtime_error("CWallet::LookupUnifiedAddress(): UFVK has no Sapling key part.");
+        }
+
+        diversifier_index_t j;
+        auto metadata = wallet.mapUfvkAddressMetadata.find(ufvkid);
+        if (metadata != wallet.mapUfvkAddressMetadata.end()) {
+            librustzcash_sapling_diversifier_index(
+                    ufvk.value().GetSaplingKey().value().dk.begin(),
+                    saplingAddr.d.begin(),
+                    j.begin());
+            auto receivers = metadata->second.GetReceivers(j);
+            if (receivers.has_value()) {
+                return ufvk.value().Address(j, receivers.value());
+            } else {
+                // If we don't know the receiver types at which the address was originally
+                // generated, we can't reconstruct the address.
+                return std::nullopt;
+            }
+        } else {
+            return std::nullopt;
```

Comment:
Under what conditions could we reach here? We should add a comment noting them, since it's not entirely obvious.

Reply:
I think this would actually require corruption; I'll change it to an assertion with a comment.